media/Jsontracer/dynamicGraduation.js

```diff
@@ -0,0 +1,126 @@
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+// This file referenced the result of
+// https://github.com/catapult-project/catapult/tree/444aba89e1c30edf348c611a9df79e2376178ba8/tracing
+
+export default function dynamicGraduation() {
+  const ruler = document.querySelector('.ruler');
+
+  if (!ruler) {
+    return;
+  }
+
+  const graduation = document.querySelector('.graduation');
+  const graduationList = document.querySelectorAll('.graduation');
+  const body = document.querySelector('body');
+  const rulerBlank = document.querySelector('.ruler-blank');
+  const staticRulerWidth = body.clientWidth - rulerBlank.clientWidth;
+
+  const setData = document.querySelector('.set-data');
+  const timeLimit = setData.dataset['timeLimit'];
+  const digit = setData.dataset['digit'];
+  const initGraduationCnt = timeLimit / 10 ** (digit - 1);
+  const staticGraduationWidth = parseInt(staticRulerWidth / initGraduationCnt);
+
+  if (graduation.offsetWidth < staticGraduationWidth - 10) {
+    removeGraduation(ruler, graduationList.length);
+  } else if (graduation.offsetWidth < staticGraduationWidth * 2) {
+    return;
+  } else {
+    addGraduation(ruler, graduationList.length);
+  }
+  updateGraduation(ruler, timeLimit);
+}
+
+function removeGraduation(ruler, cnt) {
+  for (let i = 0; i < cnt / 2; i++) {
+    const graduation = document.querySelector('.graduation');
+    ruler.removeChild(graduation);
+  }
+}
+
+function addGraduation(ruler, cnt) {
+  for (let i = 0; i < cnt; i++) {
+    const tempGraduation = document.createElement('div');
+    tempGraduation.className = 'graduation';
+
+    for (let i = 0; i < 5; i++) {
+      const tempSmallGraduation = document.createElement('div');
+      tempSmallGraduation.className = 'small-graduation';
+
+      if (i === 0) {
+        const index = document.createElement('div');
+        index.className = 'index';
+        tempSmallGraduation.append(index);
+      }
+
+      tempGraduation.append(tempSmallGraduation);
+    }
+    ruler.append(tempGraduation);
+  }
+}
+
+function updateGraduation(ruler, timeLimit) {
+  const rulerWidth = ruler.scrollWidth;
+  const allGraduation = document.querySelectorAll('.graduation');
+  let left = 0;
+
+  allGraduation.forEach(ele => {
+    ele.firstChild.firstChild.innerText = calculateGraduation(left / rulerWidth * timeLimit);
+    left += ele.offsetWidth;
+  });
+}
+
+function calculateGraduation(graduation) {
+  if (graduation >= 1000) {
+    return Math.round((graduation / 1000) * 10) / 10 + 'ms';
+  } else if (graduation >= 1) {
+    return Math.round(graduation) + 'us';
+  } else if (graduation === 0) {
+    return 0;
+  } else {
+    return Math.round(graduation * 1000 * 10) / 10 + 'ns';
```

Comment:
What is exact condition for this? `0 < graduation and graduation < 1` ?

Reply:
yes, else condition is `0 < graduation and graduation < 1`.

Reply:
Then it would be better to revise the condition statements like following.

```javascript
  if (graduation >= 1000) {
    return Math.round((graduation / 1000) * 10) / 10 + 'ms';
  } else if (graduation >= 1) {
    return Math.round(graduation) + 'us';
  } else if (graduation > 0) {
    return Math.round(graduation * 1000 * 10) / 10 + 'ns';
  } else if (graduation === 0) {
    return 0;
  } else {
    // assert here
  }
```

Reply:
graduation always greater than or equal to zero. Then it  would be better like this?
```
  if (graduation >= 1000) {
    return Math.round((graduation / 1000) * 10) / 10 + 'ms';
  } else if (graduation >= 1) {
    return Math.round(graduation) + 'us';
  } else if (graduation > 0) {
    return Math.round(graduation * 1000 * 10) / 10 + 'ns';
  } else if (graduation === 0) {
    return 0;
  }
```

Reply:
Just for insurance :D

```javascript
  if (graduation >= 1000) {
    return Math.round((graduation / 1000) * 10) / 10 + 'ms';
  } else if (graduation >= 1) {
    return Math.round(graduation) + 'us';
  } else if (graduation > 0) {
    return Math.round(graduation * 1000 * 10) / 10 + 'ns';
  } else if (graduation === 0) {
    return 0;
  } else {
    console.log('[WARNING] graduation is negative');
  }
```